lib/tbot/service/service.go

```diff
@@ -0,0 +1,187 @@
+func (s *Service[HandlerT]) Status() status.Status {
+	return s.status.Get()
+}
+
+// Get returns the service's handler. You should generally prefer to use Wait
+// because you usually only want to use the service if it's healthy.
+func (s *Service[HandlerT]) Get() HandlerT {
+	return s.handler
+}
+
+// ErrWrongStatus is returned from Service.Wait and Service.WaitForStatus when
+// the service is not in the required status.
+var ErrWrongStatus = errors.New("supervisor: service has wrong status")
+
+// Wait blocks until the service's status is Ready and then returns the handler.
+//
+// If the given context is canceled or reaches its deadline, Wait will unblock
+// immediately and return the context's error. This is useful for implementing
+// timeout/fallback behavior.
+//
+// If the service finishes running without becoming Ready (i.e. in one-shot
+// mode), ErrWrongStatus will be returned.
+func (s *Service[HandlerT]) Wait(ctx context.Context) (HandlerT, error) {
+	return s.WaitForStatus(ctx, status.Ready)
+}
+
+// WaitForStatus blocks until the service's status matches one of the given
+// statuses and then returns the handler.
+//
+// If the given context is canceled or reaches its deadline, WaitForStatus will
+// unblock immediately and return the context's error. This is useful for
+// implementing timeout/fallback behavior.
+//
+// If the service finishes running before reaching the desired status (i.e. in
+// one-shot mode), ErrWrongStatus will be returned.
+func (s *Service[HandlerT]) WaitForStatus(ctx context.Context, statuses ...status.Status) (HandlerT, error) {
+	notifCh, close := s.status.ChangeNotifications()
+	defer close()
+
+	if slices.Contains(statuses, s.status.Get()) {
+		return s.handler, nil
+	}
+
+	var zero HandlerT
+	for {
+		var finalValue, ctxDone bool
+		select {
+		case <-notifCh:
+			// New value ready.
+		case <-s.done:
+			finalValue = true
+		case <-ctx.Done():
+			ctxDone = true
+		}
+
+		switch {
+		case slices.Contains(statuses, s.status.Get()):
+			return s.handler, nil
+		case finalValue:
+			return zero, trace.Wrap(ErrWrongStatus)
+		case ctxDone:
+			return zero, trace.Wrap(ctx.Err())
+		default:
+			// Wait for the next update.
+		}
+	}
+}
+
+// These methods implement the InternalService API used by the supervisor.
+func (s *Service[HandlerT]) getName() string                     { return s.name }
+func (s *Service[HandlerT]) setStatus(status status.Status) bool { return s.status.Set(status) }
+func (s *Service[HandlerT]) runHandler(ctx context.Context, runtime *Runtime) error {
+	return s.handler.Run(ctx, runtime)
+}
+func (s *Service[HandlerT]) runOneShotHandler(ctx context.Context) error {
+	var handler any = s.handler
+	if os, ok := handler.(OneShotHandler); ok {
+		return os.OneShot(ctx)
+	}
+	return errNoOneShotHandler
+}
+func (s *Service[HandlerT]) finalize() { close(s.done) }
+func (s *Service[HandlerT]) registerToSupervisor() bool {
+	return s.registered.CompareAndSwap(false, true)
+}
```

Comment:
`registerToSupervisor` would suggest it does more than it does

```suggestion
func (s *Service[HandlerT]) markRegisteredToSupervisor() bool {
	return s.registered.CompareAndSwap(false, true)
}
```

Reply:
Agreed, that's much better!